### PR TITLE
Add crawler blog post

### PR DESCRIPTION
**Pull request by aldenstpage**
**Please refrain from linking this post anywhere (like Hacker News) until we've integrated popularity data in CC Search prod**; I'd like to have our search algorithm polished up before bringing more scrutiny to CC Search.

This adds a blog post detailing how our image crawling system works. I'm interested in feedback about structure and clarity.

### CODE REVIEW

File: content/blog/entries/crawling-500-million/contents.lr [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+- A lot of metadata will be produced by this crawler. The step of integrating it with our internal systems needs to not block resizing tasks. That suggests that a message bus will be necessary to buffer messages before they are written into our data layer, where writes can be expensive.
+- We want to have a basic idea of how the crawl is progressing in the form of summaries of error counts, status codes, and crawl rates for each source.
+
+To summarize, the challenge here isn't so much making a really fast crawler as much as it is tailoring the crawl speed to each source. At a minimum, we'll need to deal with concurrency and parallelism, provisioning and managing the life cycle of crawler infrastructure, pipelines for capturing output data, a way to monitor the progress of the crawl, a suite of tests to make sure the system behaves as expected, and a reliable way to enforce a "politeness" policy. That's not a trivial project, particularly for our tiny three person tech team (of which only one person is available to do all of the crawling work). Can't we just use an off-the-shelf open source crawler?
+
+## What about existing open source crawlers?
```
Comment (kgodey) [inline content/blog/entries/crawling-500-million/contents.lr:38]: It may be worth adding information about what open source crawlers you considered.

File: content/blog/entries/crawling-500-million/contents.lr [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+
+## Designing the crawler
+We know we're not going to be able to crawl 500 million images with one virtual machine and a single IP address, so it is obvious from the start that we are going to need a way to distribute the crawling and analysis tasks over multiple machines. A basic queue-worker architecture will do the job here; when we want to crawl an image, we can dispatch the URL to an inbound images queue, and a worker eventually pops that task out and processes it. Kafka will handle all of the hard work of partitioning and distributing the tasks between workers.
+  
+The worker processes do the actual analysis of the images, which essentially entails downloading the image, extracting interesting properties, and sticking the resulting metadata back into a Kafka for later downstream processing. The worker will also have to include some instrumentation for conforming to rate limits and error reporting.
```
Comment (kgodey) [inline content/blog/entries/crawling-500-million/contents.lr:46]: "back into a Kafka" --> "back into a Kafka topic"?

File: content/blog/entries/crawling-500-million/contents.lr [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+
+We also know that we will to share some information about crawl progress between worker processes, such as whether we've exceeded our proscribed rate limit for a website, the number of times we've seen a status code in the last minute, how many images we've processed so far, and so on. Since we're only interested in sharing application state and aggregate statistics, a lightweight key/value store like Redis seems like a good fit.
```
Comment (kgodey) [inline content/blog/entries/crawling-500-million/contents.lr:48]: "will to" --> "will need to"?

File: content/blog/entries/crawling-500-million/contents.lr [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+
+Finally, we need a maestro process that centrally controls the crawl. This key governing process will be responsible for making sure our crawler workers are behaving properly by moderating crawl rates for each source, taking action in the face of errors, and reporting statistics to the operators of the crawler. We'll call this process the crawl monitor.
```
Comment (kgodey) [inline content/blog/entries/crawling-500-million/contents.lr:50]: I haven't encountered "maestro process" before.

File: content/blog/entries/crawling-500-million/contents.lr [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+If our heuristic fails to correctly approximate the bandwidth capabilities of a site, we are going to start encountering problems. For one, we might exceed the server-side rate limit, which means we will see `429 Rate Limit Exceeded` and `403 Forbidden` errors instead of the images we're trying to crawl. Worse yet, the upstream source might continue to happily serve requests while we suck up all of their traffic capacity, resulting in other users being unable to view the images. Clearly, in either scenario, we need to either reduce our crawl rate or even give up crawling the source entirely if it appears that we are impacting their uptime.
+
+To handle these situations, we have two tools in our toolbox: a sliding window recording the status code of every request made we've made to each domain in the last 60 seconds, and a list of the last 50 status for each website. If the number of errors in our one minute window exceed 10%, something is wrong; wait a minute before trying again. If we have encountered many errors in a row, however, that suggests that we're having trouble with a particular site, so we ought to give up crawling the source and raise an alert.
```
Comment (kgodey) [inline content/blog/entries/crawling-500-million/contents.lr:98]: "last 50 status for each website" is unclear.

Comment (kgodey) [reply content/blog/entries/crawling-500-million/contents.lr:98]: This sentence was a little hard to parse for me: "If the number of errors in our one minute window exceed 10%, something is wrong; wait a minute before trying again." Perhaps something like "we should wait a minute before trying again"?

File: content/blog/entries/crawling-500-million/contents.lr [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+
+*<center>Detecting persistent crawl errors</center>*
+
+In practice, keeping a sliding window for tracking error thresholds and setting reasonable minimum and maximum crawl rates has worked well enough that the circuit breaker has never been tripped.
+
+#### Enforcing rate limits with token buckets
+
+It's one thing to set a policy for crawling; it's another thing entirely to actually enforce it. How can we coordinate our multiple crawling processes to prevent them from overstepping our rate limit?
+
+The answer is to implement a distributed token bucket system. The idea behind this is that each crawler has to obtain a token from Redis before making a request. Every second, the crawl monitor sets a variable containing the number of requests that can be made against a source. Each crawler process decrements the counter before making a request. If the decremented result is above zero, the worker is cleared to crawl. Otherwise, the rate limit has been and we should wait until a token has been obtained.
```
Comment (kgodey) [inline content/blog/entries/crawling-500-million/contents.lr:166]: "rate limit has been" --> "rate limit has been reached"?

File: content/blog/entries/crawling-500-million/contents.lr [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+open-source
+cc-search
+---
+author: aldenpage
+---
+pub_date: 2020-08-14
+---
+body:
+
+## Background
```
Comment (kgodey) [inline content/blog/entries/crawling-500-million/contents.lr:15]: It might be good to make this an h3 instead of an h2 because of the current style of the website.

Comment (kgodey) [reply content/blog/entries/crawling-500-million/contents.lr:15]: and make the other headers smaller too

File: content/blog/entries/crawling-500-million/contents.lr [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+        while not token_acquired:
+            token_acquired = await self._get_token(source)
+        return await self.client.get(url)
+
+```
+
+Meanwhile, the crawl monitor process is filling up each bucket every second.
+
+### Scheduling tasks (somewhat) intelligently
+The final gotcha in the design of our crawler is that we want to crawl every single website at the same time at its proscribed rate limit. That sounds almost tautological, like something that we should be able to take for granted after implementing all of this logic for preventing our crawler from working too quickly, but it turns out our crawler's processing capacity itself is a limited and contentious resource. We can only schedule so many tasks simultaneously on each worker, and we need to ensure that tasks from a single website aren't starving other sources of crawl capacity.
```
Comment (zackkrida) [inline content/blog/entries/crawling-500-million/contents.lr:200]: ~proscribed~ prescribed

File: content/blog/entries/crawling-500-million/contents.lr (hunk 4 above)
Comment (aldenstpage) [reply content/blog/entries/crawling-500-million/contents.lr]: I replaced this with the less flowery "supervising process"

File: content/blog/entries/crawling-500-million/contents.lr (hunk 1 above)
Comment (aldenstpage) [reply content/blog/entries/crawling-500-million/contents.lr]: I held back on naming any specific frameworks because I didn't want to accuse anyone of being in "a poorer state of maintenance than one would think at first glance"

Comment (kgodey) [reply content/blog/entries/crawling-500-million/contents.lr]: Perhaps find a more diplomatic way to put it? I think it would be useful for readers to know what limitations of existing crawlers are, even if it's just in our opinion.

Comment (aldenstpage) [reply content/blog/entries/crawling-500-million/contents.lr]: I rephrased that sentence and listed the three crawling frameworks I tried out

File: content/blog/entries/crawling-500-million/contents.lr [hunk 9]
```diff
@@ -0,0 +1,485 @@
...
+
+#### Concurrency with `asyncio`
+Crawling is a massively IO bound task. The workers need to maintain lots of simultaneous open connections with internal systems like Kafka and Redis as well as 3rd party websites holding the target images. Once we have the image in memory, performing our actual analysis task is easy and cheap. For these reasons, an asynchronous approach seems more attractive than using multiple threads of execution. Even if our image processing task grows in complexity and becomes CPU bound, we can get the best of both worlds by offloading heavyweight tasks to a process pool.
+
+Another reason that an asynchronous approach may be desirable is that we have several interlocking components which need to react to events in real-time: our crawl monitoring process needs to simultaneously control the rate limiting process and also interrupt crawling if errors are detected; our worker processes need to consume crawl events, process images, upload thumbnails, and produce metadata events. Coordinating all of these components through inter-process communication could be difficult, but breaking up tasks into small pieces and yielding to the event loop is comparatively easy.
```
[the remark this answers is not in the document]
Comment (aldenstpage) [reply content/blog/entries/crawling-500-million/contents.lr]: I'm going to assume high level familiarity with the tradeoffs between asyncio and multiprocessing. Unfortunately there aren't very many good explanations out there and it's out of scope for this article. I did add a link to the Python docs explaining why blocking code ought to be offloaded to a new process however.

**Comment by aldenstpage** [comment]
Thanks to everyone who took the time to give this a close reading and provide detailed feedback!
